blogposts/2020/optimizing-a-code-intelligence-indexer.md

```diff
@@ -0,0 +1,196 @@
+
+#### Stop looking at the stuff that doesn't matter
+
+We've greatly reduced the number of AST traversals that we perform, but that doesn't mean that the traversals we're still doing are efficient. When we traverse an AST to gather comments and construct monikers, we do so only for nodes that define a symbol (symbols that are not a definition will use the hover text and moniker of the symbol they reference).
+
+One of the inputs of the traversal is an ordered list of token positions for each definition in the file. When a node is visited, we check to see if the node encloses a position in the list. If it doesn't, the subtree can be skipped. If it does, then the node's children should be visited as the node's subtree contains a definition. If a node's position _equals_ a position in the list, then the node itself is a definition and a hover text and moniker entry should be set for that node.
+
+Because the list of positions are ordered, we can use binary search to efficiently determine which of these three conditions are true for a node. But efficient does not equate to _most efficient_. Each layer of the tree received the same list of positions, meaning that the binary search operation for every node had the same cost.
+
+It is [much more efficient](https://github.com/sourcegraph/lsif-go/pull/84) to instead pass along only the positions which can match a subtree. This change means that the binary search for a node decreases as its depth increases: the leaves of the tree have to search a much smaller list. Because the number of nodes in the tree (generally) increases with depth, this cost savings is compounded.
+
+#### Don't do lots of little work; do little lots of work
+
+An LSIF indexer is a tool that spits out an index file. The index file is generally larger than input source code due to making the implicit definition and reference relationships explicit. It stands to reason that the performance of the indexer will eventually be dependent on the performance of formatting the output and writing to disk.
+
+The LSIF index file is formatted as sequence of newline-separated JSON elements. These elements are generally pretty small with the exception of hover content vertexes and contains edges adjacent to large documents. From a dump of 14 million lines, the median line contained only 91 characters (a line with 48 characters being the smallest).
+
+Unfortunately, we were writing each line to an unbuffered file. Without a buffer, writing each element requires a [syscsall](https://man7.org/linux/man-pages/man2/write.2.html) to commit a small amount of data to the target file. With a [buffer](https://hashrocket.com/blog/posts/go-performance-observations#buffered-io), the writes are batched so that 4096 characters are written to disk at a time. This reduces the number of syscalls, and therefore number of userland/kernel context switches, by 98%.
```

Comment:
you sure you want to confess to that :-) ?

Reply:
If it stops me or another developer from doing it again in the future, then yes!